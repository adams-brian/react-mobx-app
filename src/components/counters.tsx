import * as React from 'react';
import { AppState } from '../store';
import { Counter } from './counter';
import { loading } from './loading';

export const Counters = 
loading<AppState>('appState', (appState) => appState.loadCounters())(
class extends React.Component<{appState: AppState}> {
  public render() {
    return (
      <div className="counters-container">
        <h1>Counters</h1>
        <h3>Count: {this.props.appState.countersLength}</h3>
        <div className="counter-container">
          {this.props.appState.counters.map((counter, index) => 
            (<Counter 
              key={index} /* not ideal, but it works in this case */
              index={index}
              counter={counter}
              increment={this.increment}
              decrement={this.decrement}
              reset={this.reset}
              remove={this.remove}
            />))}
          <div className="add-counter rounded" onClick={this.addCounter}>
            <span className="fa fa-plus" />
          </div>
        </div>
      </div>
    );
  }
  private increment = (index: number) =>
    this.props.appState.increment(index);
  private decrement = (index: number) =>
    this.props.appState.decrement(index);
  private reset = (index: number) =>
    this.props.appState.reset(index);
  private remove = (index: number) =>
    this.props.appState.remove(index);
  private addCounter = () =>
    this.props.appState.addCounter();
}
);