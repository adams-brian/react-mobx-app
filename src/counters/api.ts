import { debounce } from 'lodash';

export interface ICountersApi {
  loadCounters: () => Promise<number[]>;
  saveCounters: (counters: number[]) => void;
}

export default class CountersApi implements ICountersApi {

  public saveCounters: (counters: number[]) => void;

  constructor() {
    this.saveCounters = debounce(this.doSave, 1000);
  }

  public async loadCounters(): Promise<number[]> {
    const response = await fetch('http://localhost:4000/counters');
    const json = await response.json();
    return [...json.data];
  }

  private doSave(counters: number[]) {
    console.log('saving counters: ' + JSON.stringify({counters}));
    
    fetch('http://localhost:4000/counters', {
      body: JSON.stringify({counters}),
      headers: {
        'Accept': 'application/json, text/plain, */*',
        'Content-Type': 'application/json'
      },
      method: 'post',
    })
    .catch((err) => console.log('failed to save counters: ', err));
  }
}
