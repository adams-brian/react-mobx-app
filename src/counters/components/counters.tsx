import * as React from 'react';
import loading from '../../common/components/loading';
import CountersState from '../state';
import Counter from './counter';

export class Counters extends React.Component<{countersState: CountersState}> {
  public render() {
    return (
      <div className="counters-container">
        <h1>Counters</h1>
        <h3>Count: {this.props.countersState.countersLength}</h3>
        <div className="counter-container">
          {this.props.countersState.counters.map((counter, index) => 
            (<Counter 
              key={index} /* not ideal, but it works in this case */
              index={index}
              counter={counter}
              increment={this.increment}
              decrement={this.decrement}
              reset={this.reset}
              remove={this.remove}
            />))}
          <div className="add-counter rounded" onClick={this.addCounter}>
            <span className="fa fa-plus" />
          </div>
        </div>
      </div>
    );
  }
  private increment = (index: number) =>
    this.props.countersState.increment(index);
  private decrement = (index: number) =>
    this.props.countersState.decrement(index);
  private reset = (index: number) =>
    this.props.countersState.reset(index);
  private remove = (index: number) =>
    this.props.countersState.remove(index);
  private addCounter = () =>
    this.props.countersState.addCounter();
}

export default loading<CountersState, {}>('countersState', (countersState) => countersState.loadCounters())(
  Counters
);
