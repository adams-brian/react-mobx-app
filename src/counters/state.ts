import { action, computed, observable } from 'mobx';

import { ICountersApi } from './api';

export default class CountersState {
  @observable public counters: number[] = [];

  @computed get countersLength() {
    return this.counters.length;
  }

  constructor(private api: ICountersApi) { }

  @action public increment(index: number) {
    this.counters[index]++;
    this.api.saveCounters(this.counters);
  }
  @action public decrement(index: number) {
    this.counters[index]--;
    this.api.saveCounters(this.counters);
  }
  @action public reset(index: number) {
    this.counters[index] = 0;
    this.api.saveCounters(this.counters);
  }
  @action public remove(index: number) {
    this.counters.splice(index, 1);
    this.api.saveCounters(this.counters);
  }
  @action public addCounter() {
    this.counters.push(0);
    this.api.saveCounters(this.counters);
  }

  @action public async loadCounters(): Promise<void> {
    this.counters = [];
    this.counters = await this.api.loadCounters();
  }
}
