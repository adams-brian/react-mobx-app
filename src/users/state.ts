import { action, observable } from 'mobx';
import { generate } from 'shortid';

import { IUsersApi } from './api';

export interface IUser { 
  _id: string;
  firstname: string;
  lastname: string;
}

export class UsersState {
  @observable public users: IUser[] = [];

  constructor(private api: IUsersApi) { }

  @action public deleteUser(id: string) {
    this.users = this.users.filter(user => user._id !== id);
    this.api.deleteUser(id);
  }

  @action public async loadUsers(): Promise<void> {
    this.users = await this.api.loadUsers();
  }

  @action public async createUser(user: IUser) {
    const newUser = Object.assign({}, user);
    newUser._id = generate();
    this.users.push(newUser);
    this.api.createUser(newUser);
  }

  @action public updateUser(user: IUser) {
    this.users.splice(this.users.findIndex(u => u._id === user._id), 1, user);
    this.api.updateUser(user);
  }
}
