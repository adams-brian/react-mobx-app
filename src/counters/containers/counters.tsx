import { loading } from '../../common/components/loading';
import { CountersState } from '../state';

import { Counters as CountersComponent } from '../components/counters';

export const Counters = 
loading<CountersState>('countersState', (countersState) => countersState.loadCounters())(
  CountersComponent
);
