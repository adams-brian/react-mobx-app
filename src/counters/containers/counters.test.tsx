import { mount, shallow } from 'enzyme';
import * as React from 'react';
import * as sinon from 'sinon';

import { CountersState } from '../state';
import { Counters } from './counters';

let countersState: CountersState;

beforeEach(() => {
  countersState = new CountersState({
    loadCounters: sinon.fake.resolves([]),
    saveCounters: sinon.fake()
  });
  sinon.spy(countersState, 'loadCounters');
});

describe("counters", () => {
  it('renders as expected', () => {
    const component = shallow(<Counters countersState={countersState} />);
    expect(component).toMatchSnapshot();
  });
  it('calls appState.loadCounters()', () => {
    expect((countersState.loadCounters as sinon.SinonSpy).called).toBe(false);
    mount(<Counters countersState={countersState} />);
    expect((countersState.loadCounters as sinon.SinonSpy).calledOnce).toBe(true);
  });
});
