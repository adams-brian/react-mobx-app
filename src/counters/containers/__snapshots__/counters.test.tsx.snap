// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`counters renders as expected 1`] = `
<loading(Counters)
  countersState={
    CountersState {
      "api": Object {
        "loadCounters": [Function],
        "saveCounters": [Function],
      },
      "counters": Array [],
    }
  }
/>
`;
